ports/stm32f4/common-hal/busio/SPI.c

```diff
@@ -79,38 +82,80 @@ void common_hal_busio_spi_construct(busio_spi_obj_t *self,
     uint8_t sck_len = sizeof(mcu_spi_sck_list)/sizeof(*mcu_spi_sck_list);
     uint8_t mosi_len = sizeof(mcu_spi_mosi_list)/sizeof(*mcu_spi_mosi_list);
     uint8_t miso_len = sizeof(mcu_spi_miso_list)/sizeof(*mcu_spi_miso_list);
-
     bool spi_taken = false;
-    //sck
+
+    //SCK is not optional. MOSI and MISO are
     for(uint i=0; i<sck_len;i++) {
         if (mcu_spi_sck_list[i].pin == sck) {
-            //mosi
-            for(uint j=0; j<mosi_len;j++) {
-                if (mcu_spi_mosi_list[j].pin == mosi) {
-                    //miso
-                    for(uint k=0; k<miso_len;k++) {
-                        if ((mcu_spi_miso_list[k].pin == miso) //everything needs the same index
-                            && (mcu_spi_sck_list[i].spi_index == mcu_spi_mosi_list[j].spi_index)
-                            && (mcu_spi_sck_list[i].spi_index == mcu_spi_miso_list[k].spi_index)) {
-                            //keep looking if the SPI is taken, edge case
-                            if(reserved_spi[mcu_spi_sck_list[i].spi_index-1]) {
-                                spi_taken = true;
-                                continue;
+            //if both MOSI and MISO exist, loop search normally
```

Comment:
It's tempting to suggest structuring the search like this:
```
for(SCK alternatives) {
   if(mosi != mp_const_none) search for mosi and continue outer loop if unsuccessful;
   if(miso != mp_const_none) search for miso and continue outer loop if unsuccessful;
}
```
which eliminates the duplicated code and removes a level of loop nesting.  However, as this is not performance critical code, doing it what you felt was the most straightforward way possible also has merit.

Reply:
@jepler Unfortunately, this approach does not function properly in the both-enabled case, as it excludes an edge case where there is a second SPI peripheral on the MOSI pin that is the only viable match to MISO. The only way to properly find that is to have the loops nested. 

I think I could still fit it all in one nested loop, but it'd take so many flags and crazy conditionals that I'm not convinced the marginal code lines reduction would be worth the cost in readability.

Reply:
Well, without a `goto:` anyway. Dan's link above reminded me. I try not to use them myself.

Reply:
Here's a goto version that covers all edge cases, I think: 
https://gist.github.com/hierophect/8d304e7d44c9816e294874529937019d

As I suspected, it only saves about 15 lines of code, and I think it's harder to decipher what's going on. I could be missing an easier opportunity though.

Reply:
eek, `goto` a label inside a loop? The compiler may have to de-optimize things to allow this to even work. Interesting coding exercise, but your old code is ok by me. :)

Reply:
@dhalbert lol it was really more an exercise in convincing myself it was a bad idea to rework ;)

Reply:
UART has the same issue show up so I think it's a good problem to have thought hard about.